File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution(object):
    def getMaximumGold(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        checked = set()
        def pather(x,y):
            potential = [0]
            if grid[x][y] != 0:
                if x > 0 and (x-1,y) not in checked:
                    checked.add((x-1,y))
                    potential.append(pather(x-1,y))
                    checked.remove((x-1,y))
                if x < len(grid)-1 and (x+1,y) not in checked:
                    checked.add((x+1,y))
                    potential.append(pather(x+1,y))
                    checked.remove((x+1,y))
                if y > 0 and (x,y-1) not in checked:
                    checked.add((x,y-1))
                    potential.append(pather(x, y-1))
                    checked.remove((x,y-1))
                if y < len(grid[x]) - 1 and (x, y+1) not in checked:
                    checked.add((x, y+1))
                    potential.append(pather(x, y+1))
                    checked.remove((x,y+1))
            return grid[x][y] + max(potential)
        maxer = 0
        for x in range(len(grid)):
            for y in range(len(grid[x])):
                if grid[x][y] != 0:
                    checked.add((x,y))
                    value = pather(x,y)
                    maxer = max(maxer, value)
                    checked.remove((x,y))
        return maxer

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.java
class Solution {
    public int getMaximumGold(int[][] grid) {
        int answer = 0;
        for(int x = 0; x< grid.length; x++){
            for(int y = 0; y < grid[x].length; y++){
                answer = Math.max(answer, gold(x,y,grid));
            }
        }
        
        return answer;
    }
    
    public int gold(int x, int y, int[][] grid){
        int value = grid[x][y];
        int answer = 0;
        grid[x][y] = 0;
        if(x > 0 && grid[x-1][y] != 0){
            answer = Math.max(answer, gold(x-1, y, grid));
        }
        if(x < grid.length - 1 && grid[x+1][y] != 0){
            answer = Math.max(answer, gold(x+1, y, grid));
        }
        
        if(y > 0 && grid[x][y-1] != 0){
            answer = Math.max(answer, gold(x, y-1, grid));
        }
        if(y < grid[0].length - 1 && grid[x][y+1] != 0){
            answer = Math.max(answer, gold(x, y+1, grid));
        }
        
        grid[x][y] = value;
        
        return value + answer;
        
    }
}